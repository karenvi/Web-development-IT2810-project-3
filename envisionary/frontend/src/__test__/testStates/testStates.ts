import { useEffect } from "react";
import { useRecoilValue } from "recoil";

export const RecoilObserver = ({node, onInput}: any) => {
  const value = useRecoilValue(node);
  useEffect(() => onInput(value), [onInput, value]);
  return null;
};