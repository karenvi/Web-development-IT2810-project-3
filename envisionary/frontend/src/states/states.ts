import { atom } from "recoil";

export const searchQueryState = atom({
  key: 'searchQueryID',
  default: '',
});

// Makes "country" the default state to search for
export const categoryState = atom({
  key: 'categoryStateID',
  default: 'Country',
});


// Kan fjerne denne staten, men er greit å ha i begynnelsen så man skjønner hovedprinsippet til recoil!
export const testRecoilState = atom({
    key: 'testRecoilID',
    default: 1,
  });