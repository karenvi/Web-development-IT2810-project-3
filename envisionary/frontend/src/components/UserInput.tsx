import Box from '@mui/material/Box';
import TextField from '@mui/material/TextField';
import InputLabel from '@mui/material/InputLabel';
import MenuItem from '@mui/material/MenuItem';
import FormControl from '@mui/material/FormControl';
import Select from '@mui/material/Select';
import { useRecoilState } from 'recoil';
import { categoryState, pageState, searchQueryState } from '../states/states';
import { useNavigate } from 'react-router-dom';

// This component takes in a search query from user and what category the user has picked to search in.
function UserInput() {
  const [category, setCategory] = useRecoilState(categoryState);
  const [searchQuery, setSearchQuery] = useRecoilState(searchQueryState);
  const [page, setPage] = useRecoilState(pageState);
  const navigate = useNavigate();

  const labelSearch = "Search after " + category.toLowerCase();

  // To ensure that the SPA requirement is achieved in the search (doesnt change url on search)
  async function onSubmit(e: React.FormEvent) {
    navigate(``, { replace: true });
    e.preventDefault();
  }
  

  return (
      <Box
        sx={{
          m: 5, p: '35px', width: '40%', maxWidth: '450px', minWidth: '330px', backgroundColor: 'white', display: 'flex', justifyContent: 'center',
          boxShadow: '0px 2px 1px -1px rgb(0 0 0 / 20%), 0px 1px 1px 0px rgb(0 0 0 / 14%), 0px 1px 3px 0px rgb(0 0 0 / 12%)',
          borderRadius: '10px'
        }}
      >
        <form action="/" method="get" autoComplete="off" onSubmit={onSubmit}>
          <label htmlFor="header-search">
            <span className="visually-hidden">Search by {category.toLowerCase()}</span>
          </label>
          <TextField
            label={labelSearch}
            variant="outlined"
            type="text"
            id="header-search"
            name="s"
            value={searchQuery}
            onInput={(event) => {
              setSearchQuery((event.target as HTMLInputElement).value);
              setPage(0);
            }}
          />
        </form>
        <FormControl fullWidth sx={{ width: '150px', ml: "10px" }}>
          <InputLabel id="demo-simple-select-label">Category:</InputLabel>
          <label htmlFor="header-search">
            <span className="visually-hidden">Select what to search category to search in</span>
          </label>
          <Select
            labelId="demo-simple-select-label"
            id="demo-simple-select"
            value={category}
            label="Category:"
            onChange={(event) => { setCategory(event.target.value as string)}}
          >
            <MenuItem value='Country'>Country</MenuItem>
            <MenuItem value='Continent'>Continent</MenuItem>
          </Select>
        </FormControl>
      </Box>
  )
}
export default UserInput;