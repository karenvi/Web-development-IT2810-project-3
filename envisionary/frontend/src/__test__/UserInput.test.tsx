import "@testing-library/jest-dom";
import { act, fireEvent, render, screen } from "@testing-library/react";
import renderer from 'react-test-renderer';
import UserInput from "../components/UserInput";
import { RecoilRoot } from "recoil";
import { BrowserRouter as Router } from "react-router-dom";
import { RecoilObserver } from "./testStates/testStates";
import { searchQueryState } from "../states/states";
export {};


/* const mocks = [
    {   
        request: {
            query: GET_COUNTRIES,
            variables: {},
        },
        result: {
            data: {
                countries: { id: "1", Rank: "36", Country: "Afghanistan", Capital: "Kabul", 
                Continent: "Asia",  Population2022: "41128771", Population2020: "38972230", Population2015: "33753499",
                Population2010: "28189672", Population2000: "19542982", Population1990: "10694796", Population1980: "12486631",
                Population1970: "10752971", Area: "652230", Density: "63.0587", GrowthRate: "1.0257", WorldPopulationPercentage: "0.52",   
                Reviews: [{Name: "Trond", ReviewText: "Flott land Ã¥ feriere i.", ReviewDate: "2022-10-26T13:50:35.783Z", Rating: 3.5},
                        {Name: "", ReviewDate: "", Date: "", Rating: 3.5}] }
            }
        }
    }
]; */

const data: Array<any> = [
    { _id: "1", Area: "120538", CCA3: "PRK", Capital: "Pyongyang", Continent: "Asia", Country: "North Korea", Density: "216.2755", 
    GrowthRate: "1.0038", Population1970: "14996879", Population1980: "17973650", Population1990: "20799523", Population2000: "23367059",
    Population2010: "24686435", Population2015: "25258015", Population2020: "25867467", Population2022: "26069416", Rank: "56", Reviews: null,
    WorldPopulationPercentage: "0.33" },
                            
    { _id: "2", Area: "25713", CCA3: "MKD", Capital: "Skopje", Continent: "Europe", Country: "North Macedonia", Density: "81.4218", GrowthRate: "0.9954",
    Population1970: "1656783", Population1980: "1907023", Population1990: "2044174", Population2000: "2037936", Population2010: "2093828",
    Population2015: "2107962", Population2020: "2111072", Population2022: "2093599", Rank: "150", Reviews: null, WorldPopulationPercentage: "0.03"},
    
    { _id: "3", Area: "464", CCA3: "NFK", Capital: "Saipan", Continent: "Oceania", Country: "Northern Mariana Islands", Density: "106.7909", 
    GrowthRate: "1.0014", Population1970: "10143", Population1980: "17613", Population1990: "48002", Population2000: "80338", Population2010: "54087",
    Population2015: "51514", Population2020: "49587", Population2022: "49551", Rank: "210", Reviews: null, WorldPopulationPercentage: "0"},
];


describe("Testing UserInput component", () => {
    
    it("snapshot test", () => {

        const component = renderer.create(
            <Router>
                <RecoilRoot>
                    <UserInput queryFilteredCountries={data} />
                </RecoilRoot>
            </Router>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("result is changed when typing", () => {

        const onInput = jest.fn(); 

        render(
            <Router>
                <RecoilRoot>
                    <RecoilObserver node={searchQueryState} onInput={onInput} />
                    <UserInput queryFilteredCountries={data} />
                </RecoilRoot>,
            </Router>
          );

        const component = screen.getByTestId("header-search"); 
        
        fireEvent.change(component, {target: {value: 'North Korea'}});
    });


   
   /*beforeEach(() => {
        component = create(<UserInput queryFilteredCountries={data} />);
    });
 */


});



/* it("renders without error", async () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Countries />
      </MockedProvider>,
    );
    expect(await screen.findByText("Loading...")).toBeInTheDocument();
}); */