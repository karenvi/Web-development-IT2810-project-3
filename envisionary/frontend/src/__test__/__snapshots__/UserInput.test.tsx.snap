// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing UserInput component snapshot test 1`] = `
<div
  className="MuiBox-root css-1mwmua6"
>
  <form
    action="/"
    autoComplete="off"
    method="get"
    onSubmit={[Function]}
  >
    <label
      htmlFor="header-search"
    >
      <span
        className="visually-hidden"
      >
        Search by 
        country
      </span>
    </label>
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      onInput={[Function]}
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor="header-search"
        id="header-search-label"
      >
        Search for country
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          data-testid="search-test"
          disabled={false}
          id="header-search"
          name="s"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-yjsfm1"
          >
            <span>
              Search for country
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </form>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth css-1tp1e4i-MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      id="demo-simple-select-label"
    >
      Category:
    </label>
    <label
      htmlFor="header-search"
    >
      <span
        className="visually-hidden"
      >
        Select what to search category to search in
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      onClick={[Function]}
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="demo-simple-select-label demo-simple-select"
        className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="demo-simple-select"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        Country
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        data-testid="select-test"
        disabled={false}
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value="Country"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-14lo706"
        >
          <span>
            Category:
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
